go/vt/vtgate/semantics/typer.go

```diff
@@ -20,17 +20,18 @@ import (
 	"vitess.io/vitess/go/sqltypes"
 	querypb "vitess.io/vitess/go/vt/proto/query"
 	"vitess.io/vitess/go/vt/sqlparser"
+	"vitess.io/vitess/go/vt/vtgate/engine"
 )
 
+// typer is responsible for setting the type for expressions
+// it does it's work after visiting the children (up), since the children types is often needed to type a node.
 type typer struct {
```

Comment:
vtgate engine should have a dependency on semantics. But we can change this if we get into circular dependency.